torch/onnx/utils.py

```diff
@@ -147,11 +147,20 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
     if export_params:
         # NB: OrderedDict values is not actually a list, but trace.export is
         # not duck-typed and expects an actual list.
-        proto = trace.export(list(_unique_state_dict(model).values()), _onnx_opset_version)
+        proto, export_map = trace.export(list(_unique_state_dict(model).values()), _onnx_opset_version, export_zip)
     else:
-        proto = trace.export([], _onnx_opset_version)
+        proto, export_map = trace.export([], _onnx_opset_version)
+
+    if export_zip:
+        import zipfile
+        compression = zipfile.ZIP_DEFLATED if compress_zip else zipfile.ZIP_STORED
+        with zipfile.ZipFile(f, 'w', compression=compression) as z:
+            z.writestr('__GRAPH_PROTO', proto)
```

Comment:
constant for it? and it's model proto btw

Reply:
another crazy ideas is that zip files can have arbitrary crap in the beginning of the file:

```
$ unzip -l foo.zip
Archive:  foo.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        6  03-23-2018 15:48   a
        7  03-23-2018 15:48   b
---------                     -------
       13                     2 files

$ (echo "crap, this is random crap" && cat foo.zip) > bar.zip

unzip -l bar.zip
Archive:  bar.zip
warning [bar.zip]:  26 extra bytes at beginning or within zipfile
  (attempting to process anyway)
  Length      Date    Time    Name
---------  ---------- -----   ----
        6  03-23-2018 15:48   a
        7  03-23-2018 15:48   b
---------                     -------
       13                     2 files

$ ls -la *.zip
-rw-r--r--. 1 dzhulgakov users 321 Mar 23 15:51 bar.zip
-rw-r--r--. 1 dzhulgakov users 295 Mar 23 15:49 foo.zip
```

So we could just prepend the onnx proto as the first thing so deserializing as proto message directly still works. Not sure it's a good idea though :)

Reply:
Hmm, that's intriguing. Do you think we would ever end up with multiple ModelProto's in an archive? that would be the only thing stopping me from doing this